protos/flyteidl/datacatalog/datacatalog.proto

```diff
@@ -33,6 +34,28 @@ service DataCatalog {
 
     // Return a paginated list of datasets
     rpc ListDatasets (ListDatasetsRequest) returns (ListDatasetsResponse);
+
+    // Get an artifact and the corresponding data. If the artifact does not exist,
+    // try to reserve a spot for populating the artifact.
+    //
+    // Once you preserve a spot, you should call ExtendReservation API periodically
+    // to extend the reservation. Otherwise, the reservation can expire and other
+    // tasks may take the spot.
+    //
+    // Note: We may have multiple concurrent tasks with the same signature
+    // and the same input that try to populate the same artifact at the same time.
+    // Thus with reservation, only one task can run at a time, until the reservation
+    // expire.
+    //
+    // Note: If the task A does not extend the reservation in time and the reservation
+    // may expire. Another task B may take over the reservation and now we have two tasks
+    // A and B running in parallel. So a third task C may get the Artifact from A or B,
+    // whoever is the last writer.
+    rpc GetOrReserveArtifact (GetOrReserveArtifactRequest) returns (GetOrReserveArtifactResponse);
+
+    // Extend the reservation to keep it from expire. If the reservation expire,
+    // other tasks can take over the reservation by calling GetOrReserveArtifact.
+    rpc ExtendReservation (ExtendReservationRequest) returns (ExtendReservationResponse);
```

Comment:
Should we also add a ReleaseReservation() api? for now the implementation may just return success... but it'll allow the execution engine (propeller) to "cleanup" reservations if the task fails... (optimistically).. even if it fails to do so, the reservation will still expire.. so it's not a correctness fix but more of an optimization so other tasks can start sooner...

CC @kumare3 
```suggestion
    rpc ExtendReservation (ExtendReservationRequest) returns (ExtendReservationResponse);
    
    rpc ReleaseReservation (ReleaseReservationRequest) returns (ReleaseReservationResponse);
```

Reply:
Yeah I like the idea.

Reply:
added